coalahtml/tree.py

```diff
@@ -0,0 +1,65 @@
+from collections import OrderedDict
+from queue import Queue
+import os
+
+
+class Node:
+
+    def __init__(self, path):
+        self.child = {}
+        self.parent = None
+        self.path = path
+        self.abspath = ""
+
+
+class Tree:
+
+    def __init__(self, file_list):
+        """
+        Create a root and build the graph.
+        """
+        self.root = Node(Tree.findroot(file_list))
+        self.root.abspath = self.root.path
+        self.fl_split = [path.split(os.sep) for path in file_list]
+        self._build(self.fl_split)
+
+    @staticmethod
+    def findroot(file_list):
+        prefix = os.path.commonprefix(file_list)
+        last_sep = prefix.rfind(os.sep)
+        return prefix[:last_sep]
+
+    def _add_node(self, new_node, prev_node, head):
+        if prev_node.child.get(new_node.path, None) is None:
+            prev_node.child[new_node.path] = new_node
+        else:
+            # Duplicates
+            return prev_node.child[new_node.path]
+
+        return new_node
+
+    def _build(self, fl_split):
+        start = len(self.root.path.split(os.sep))
+        for each_path in fl_split:
+            prev_node = self.root
+            for item in each_path[start:]:
+                new_node = Node(item)
+                new_node.parent = prev_node
+                new_node.abspath = os.path.join(prev_node.abspath,
+                                                new_node.path)
+                prev_node = self._add_node(new_node, prev_node, self.root)
+
+    def extract(self):
```

Comment:
​Can we specify that in docstring alone? ​Do you want to expand the name
too? Suggestions?
ᐧ

On Thu, May 26, 2016 at 11:53 PM, Attila <notifications@github.com> wrote:

> In coalahtml/tree.py
> <https://github.com/coala-analyzer/coala-html/pull/46#discussion_r64795442>
> :
>
> > +            return prev_node.child[new_node.path]
> > +
> > +        return new_node
> > +
> > +    def _build(self, fl_split):
> > +        start = len(self.root.path.split(os.sep))
> > +        for each_path in fl_split:
> > +            prev_node = self.root
> > +            for item in each_path[start:]:
> > +                new_node = Node(item)
> > +                new_node.parent = prev_node
> > +                new_node.abspath = os.path.join(prev_node.abspath,
> > +                                                new_node.path)
> > +                prev_node = self._add_node(new_node, prev_node, self.root)
> > +
> > +    def extract(self):
>
> From tree you can extract a lot of things. You should specify what exactly
> are you extracting here
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/coala-analyzer/coala-html/pull/46/files/8fc09d0dc5c2bd4bf2032aa0657b5a63af442363..2371e9ca763b5a8791588526ebd237d580b60c44#r64795442>
>